ragthoven/__init__.py

```diff
@@ -136,6 +139,150 @@ def __init__(
         else:
             self.output_write = output_write
 
+    def _instantiate_tools(self, tool_configs):
+        """
+        Instantiate tools defined in config.iterative.tools with dependency injection.
+        """
+        tools = {}
+        if tool_configs is None:
+            return tools
+
+        available_deps = {
+            "prompt_executor": self.pexecutor,
+            "embedder": self.embedder,
+            "reranker": self.reranker,
+        }
+
+        for cfg in tool_configs:
+            model_override = None
+            cls = None
+
+            if isinstance(cfg, dict):
+                name = cfg.get("name")
+                model_override = cfg.get("model")
+            elif inspect.isclass(cfg):
+                cls = cfg
+                # Normalize to the dotted path expected by get_class
+                name = (
+                    f"{cls.__module__.split('ragthoven.tools.', 1)[-1]}.{cls.__name__}"
+                )
+            else:
+                name = cfg
+
+            if name is None:
+                raise ValueError("Tool configuration is missing 'name'")
+
+            if "." not in name:
+                raise ValueError(
+                    "Tool name must include its module prefix, e.g. 'reasoning_tools.Calculator'"
+                )
+
+            if cls is None:
+                cls = get_class("ragthoven.tools", name)
+                class_name = get_class_func_name_only(name)
+            else:
+                class_name = cls.__name__
+
+            requires = getattr(cls, "requires", [])
+            kwargs = {}
+            for dep in requires:
+                if dep not in available_deps:
+                    raise ValueError(f"Unsupported dependency requested: {dep}")
+                kwargs[dep] = available_deps[dep]
+            if model_override is not None:
+                kwargs["model_override"] = model_override
+
+            tools[class_name] = cls(**kwargs)
+
+        return tools
+
+    def _get_tool_schemas(self, tools: dict):
+        return [tool.get_json_schema() for tool in tools.values()]
+
+    def _execute_tool(self, tool_call, tools: dict):
+        try:
+            tool = tools[tool_call.function.name]
+        except KeyError:
+            return f"ERROR: Unknown tool '{tool_call.function.name}'"
+
```

Comment:
Consider using the existing `logger` instead of `print()` for consistency with the rest of the codebase.